### ISSUE TITLE
`RegExpRouter` fails when path regexes include parens

### ISSUE THREAD

**Issue by jmandel** [honojs/hono#1537]
### What version of Hono are you using?

3.7.3

### What runtime/platform is your app running on?

Deno

### What steps can reproduce the bug?

```
const app = new Hono({
  router: new TrieRouter(),
})
app.get("/dyn/:dyn/:service{(a|b)}", (c) => c.json(c.req.param()))
```

The routing behavior is different depending on whether I choose a `TrieRouter` (provides the expected result, working for paths like `/dyn/val/a`) vs `LinearRouter` and others (which lead to an `Internal Server Error`). The specific issue here is about `()` inside the `{}` of a param, but my more general question: are the different router types expected to have different semantics? The docs are all about performance characteristics, rather than any differences in behavior.

### What is the expected behavior?

If different routers are expected to produce different results, this is important to call out in the docs.

If all routers are expected to produce identical results, this would be worth ensuring in the test suite.

**Comment by NicoPlyley** [comment]
@jmandel I just tried your code with deno and could not replicate the issue. It works on both the routers for me. Can you share your entire code that is not working so I can see if there is anything additional.

**Comment by yusukebe** [comment]
Hi @jmandel thanks for raising the issue.

I'm getting an error with `RegExpRouter` instead of `TrieRouter`.

```ts
const app = new Hono({
  router: new RegExpRouter(),
})
```

<img width="1016" alt="Screenshot 2023-10-04 at 8 11 11" src="https://github.com/honojs/hono/assets/10682/d875de68-5cc6-4206-b658-0a5feae955e8">

This is not expected behavior and should be corrected or throw an `UnsupportedPathError` and fall back to the `TrieRouter`.

@usualoma :

Could you look at this issue when you have time?

**Comment by jmandel** [comment]
@yusukebe your result matches mine -- `TrieRouter` worked, and the other routers led to an error.

**Comment by yusukebe** [comment]
@jmandel 

I think `LinearRouter` and `PatternRouter` don't throw the error.

Just to confirm, the application is following code:

```ts
import { Hono } from 'hono'
import { LinearRouter } from 'hono/router/linear-router'

const app = new Hono({
  router: new LinearRouter(),
})

app.get('/dyn/:dyn/:service{(a|b)}', (c) => c.json(c.req.param()))

export app
```

And access there:

```txt
http://127.0.0.1:8787/dyn/abc/a
```

It will return:

```txt
{"dyn":"abc","service":"a"}
```

**Comment by jmandel** [comment]
```ts
import { Hono, Env, LinearRouter, RegExpRouter, TrieRouter, PatternRouter } from "https://deno.land/x/hono@v3.7.3/mod.ts";

const app = new Hono();

[LinearRouter, RegExpRouter, TrieRouter, PatternRouter]
  .map(rc => new rc())
  .map(router => new Hono({router}))
  .forEach(r => {
    r.get('/begin/:opt{(optional|skippable)}/end', (c) => c.json(c.req.param()))
    app.route(r.router.name, r)
  });

Deno.serve(app.fetch)
```

```sh
$ curl http://localhost:8000/LinearRouter/begin/optional/end
Internal Server Error

$ curl http://localhost:8000/TrieRouter/begin/optional/end
Internal Server Error

$ curl http://localhost:8000/RegExpRouter/begin/optional/end
Internal Server Error

$ curl http://localhost:8000/PatternRouter/begin/optional/end
{"opt":"optional"}
```

**Comment by yusukebe** [comment]
@jmandel 

I'm not sure if only the `PatternRouter` is successful, but in this case it will always use RegExpRouter.

Only `RegExpRouter` fails in the following test.

```ts
import { Hono } from 'hono'
import { RegExpRouter } from 'hono/router/reg-exp-router'
import { TrieRouter } from 'hono/router/trie-router'
import { PatternRouter } from 'hono/router/pattern-router'
import { LinearRouter } from 'hono/router/linear-router'

const routerClasses = [RegExpRouter, PatternRouter, TrieRouter, LinearRouter]

routerClasses.map(async (RouterClass) => {
  const router = new RouterClass<any>()
  const app = new Hono({ router })
  app.get('/begin/:opt{(a|b)}/end', (c) => {
    return c.text(`${app.routerName} - ${c.req.path}`)
  })
  try {
    const res = await app.request('/begin/a/end')
    console.log(await res.text())
  } catch (e) {
    console.log(`${app.routerName} - ${e}`)
  }
})
```

```txt
RegExpRouter - TypeError: undefined is not an object (evaluating 'matcher[1][index]')
PatternRouter - /begin/a/end
TrieRouter - /begin/a/end
LinearRouter - /begin/a/end
```

@usualoma 
This may be a bug in RegExpRouter, can you please take a look at it? I think it would be good to throw an `UnsupportedPathError`.

**Comment by usualoma** [comment]
@yusukebe 

I think this is a bad behavior of `RegExpRouter` too.
Since `RegExpRouter` uses regular expression capture groups internally, using groups in patterns, we encounter unexpected errors.

### Workaround

With "Non-capturing group" we can work around this error even in the current version.

`/begin/:opt{(?:a|b)}/end`

### Improvements on the `RegExpRouter`

There are ways to improve the following
* throw an `UnsupportedPathError`
* convert `(a|b)` to `(?:a|b)` automatically

Please wait a moment as I consider which say is best suited.

**Comment by yusukebe** [comment]
@usualoma 

Thanks! I'll comment on #1556

**Comment by yusukebe** [comment]
This was fixed with #1556.
